docs/tutorial_basics.rst

```diff
@@ -427,33 +435,35 @@ Here is the corresponding ReFrame test, where the new concepts introduced are hi
 
 .. literalinclude:: ../tutorials/basics/hellomp/hellomp1.py
    :lines: 6-
-   :emphasize-lines: 11-13
-
-
-In order to compile applications using ``std::thread`` with GCC and Clang, the ``-pthread`` option has to be passed to the compiler.
-Since the above option might not be valid for other compilers, we use pipeline hooks to differentiate based on the programming environment as follows:
+   :emphasize-lines: 10-10, 13-18
 
-.. code-block:: python
-
-   @rfm.run_before('compile')
-   def set_threading_flags(self):
-       environ = self.current_environ.name
-       if environ in {'clang', 'gnu'}:
-           self.build_system.cxxflags += ['-pthread']
 
+ReFrame delegates the compilation of a test to a :attr:`~reframe.core.pipeline.RegressionTest.build_system`, which is an abstraction of the steps needed to compile the test.
+Build systems take also care of interactions with the programming environment if necessary.
+Compilation flags are a property of the build system.
+If not explicitly specified, ReFrame will try to pick the correct build system (e.g., CMake, Autotools etc.) by inspecting the test resources, but in cases as the one presented here where we need to set the compilation flags, we need to specify a build system explicitly.
+In this example, we instruct ReFrame to compile a single source file using the ``-std=c++11 -pthread -Wall`` compilation flags.
+However, the flag ``-pthread`` is only needed to compile applications using ``std::thread`` with the GCC and Clang compilers.
+Hence, since this flag may not be valid for other compilers, we need to include it only in the tests that use either GCC or Clang.
+Similarly to the ``lang`` parameter in the previous example, the information regarding which compiler is being used is only available after the class is instantiated, so we also defer the addition of this optional compiler flag with a pipeline hook.
```

Comment:
This is not 100% accurate. The "information regarding which compiler is being used" is available after the setup stage, so setting the flags in `post-init` wouldn't work.

Reply:
Added the missing info here.